torchchat/model.py

```diff
@@ -210,18 +218,36 @@ class Transformer(nn.Module):
     def __init__(self, config: TransformerArgs) -> None:
         super().__init__()
         self.config = config
+        layers_per_stage = config.n_layers // config.n_stages
 
-        self.tok_embeddings = nn.Embedding(config.vocab_size, config.dim)
-        self.layers = nn.ModuleList(
-            TransformerBlock(config) for _ in range(config.n_layers)
+        self.tok_embeddings = (
+            nn.Embedding(config.vocab_size, config.dim)
+            if config.stage_idx == 0 else None
+        )
+
+        # Use ModuleDict so that each layer can be assigned its layer ID in the original model
+        self.layers = nn.ModuleDict()
+
+        for layer_id in range(
+            layers_per_stage * config.stage_idx, layers_per_stage * (config.stage_idx + 1)
+        ):
+            self.layers[str(layer_id)] = TransformerBlock(config)
+
+        self.norm = (
```

Comment:
minor style nit - I feel like this is more readable expressed as a default and if true modify block ala:
~~~
# only last stage handles final norm and output:
self.norm = None
self.output = None
if config.stage_idx == config.n_stages-1:
     self.norm = RMSNorm()
     self.output = nn.linear()
~~~
rather than each having identical ternary if statement at the end of the init call.

Reply:
Makes sense. Modified.